File: solutions/scala/027-search-in-rotated-sorted-array.scala
/**
 * Problem 27: Search in Rotated Sorted Array (LeetCode 33)
 * Difficulty: Med
 * Language: Scala
 */
def search(nums: Array[Int], target: Int): Int = {
  @annotation.tailrec
  def go(lo: Int, hi: Int): Int = {
    if (lo > hi) -1
    else {
      val mid = (lo + hi) / 2
      if (nums(mid) == target) mid
      else if (nums(lo) <= nums(mid)) {
        if (target >= nums(lo) && target < nums(mid)) go(lo, mid - 1)
        else go(mid + 1, hi)
      } else {
        if (target > nums(mid) && target <= nums(hi)) go(mid + 1, hi)
        else go(lo, mid - 1)
      }
    }
  }
  go(0, nums.length - 1)
}

File: solutions/rust/027-search-in-rotated-sorted-array.rs
/**
 * Problem 27: Search in Rotated Sorted Array (LeetCode 33)
 * Difficulty: Med
 * Language: Rust
 */
pub fn search(nums: Vec<i32>, target: i32) -> i32 {
    fn go(nums: &[i32], target: i32, lo: usize, hi: isize) -> i32 {
        if lo as isize > hi { return -1; }
        let mid = (lo as isize + hi) as usize / 2;
        if nums[mid] == target { return mid as i32; }
        if nums[lo] <= nums[mid] {
            if target >= nums[lo] && target < nums[mid] {
                go(nums, target, lo, mid as isize - 1)
            } else {
                go(nums, target, mid + 1, hi)
            }
        } else if target > nums[mid] && target <= nums[hi as usize] {
            go(nums, target, mid + 1, hi)
        } else {
            go(nums, target, lo, mid as isize - 1)
        }
    }
    if nums.is_empty() { return -1; }
    go(&nums, target, 0, nums.len() as isize - 1)
}
